Index: kicad.bzr/pcbnew/pcbnew_scripted.cpp
===================================================================
--- kicad.bzr.orig/pcbnew/pcbnew_scripted.cpp	2011-03-12 21:32:36.000000000 -0300
+++ kicad.bzr/pcbnew/pcbnew_scripted.cpp	2011-03-12 21:50:13.000000000 -0300
@@ -55,6 +55,7 @@
      { wxCMD_LINE_OPTION, wxT("ps-pads-drill-opt"), wxT("ps-pads-drill-opt"), wxT("Postscript pads drill option [none|small|real] (default:small)"),
           wxCMD_LINE_VAL_STRING },
      { wxCMD_LINE_SWITCH, wxT("mirror"), wxT("mirror"), wxT("mirror plot (HPGL and Postscript only)") },
+     { wxCMD_LINE_SWITCH, wxT("e"), wxT("exclude-pcb-edge"), wxT("exclude PCB edge (Gerber only)") },
      { wxCMD_LINE_SWITCH, wxT("fill-all-zones"), wxT("fill-all-zones"), wxT("fill zones before plotting") },
      { wxCMD_LINE_SWITCH, wxT("drc"), wxT("drc"), wxT("generates a design rule check report (.rpt)") },
      { wxCMD_LINE_SWITCH, wxT("svg"), wxT("svg"), wxT("plots the board in SVG format") },
@@ -223,6 +224,10 @@
         if ( parser.Found( wxT("mirror") ) )
             g_pcb_plot_options.PlotOrient = PLOT_MIROIR;
 
+        // --exclude-pcb-edge
+        if ( parser.Found( wxT("exclude-pcb-edge") ) )
+            g_pcb_plot_options.Exclude_Edges_Pcb = true;
+
         if ( parser.Found( wxT("fill-all-zones") ) )
             frame->Fill_All_Zones( false /* verbose */ );
 
